# Lesson 6 Class Exercises: More Practice

We will use today's class exercise to reinforce the learning we've done to this point

## Get Started
Import the Numpy and Pandas packages

In [1]:
import numpy as np
import pandas as pd

## Exercise 1

Read in the apple data found in the `data/apple_data.tsv` directory.

In [2]:
apple = pd.read_csv('data/apple_data.tsv', delimiter='\t')
apple.head(5)

State    1982    1987    1992    1997    2002    2007     1982  \
0        Alabama    899     626     474     479     265     231    1,548    
1         Alaska      --      --      2       4       9      10        --   
2        Arizona    214     224     235     158     185     496    1,969    
3       Arkansas    512     451     247     238     164     153    2,276    
4     California  2,904   2,787   3,082   3,454   2,120   2,074   29,933    

      1987   1992.1   1997.1   2002.1   2007.1 2002.2 2007.2  2002.3  2007.3  
0   1,217    1,011      968      594      393     191    144     386     307  
1       --      (D)       2       13       15       8     10     (D)     (D)  
2   4,881    5,770    4,003    1,581    1,344     118    424   1,459   1,249  
3   2,120    1,430    1,048      658      287     123    116     560     220  
4  34,105   42,024   49,704   38,268   22,184   1,850  1,850  35,856  20,954

Explore the data using the approach we've used thus far:
- List all of the column names
- Identify the data types
- Find the shape of the data
- Explore categorical values -- do we have the expected number of values for each?
- Get descriptive statistics for the numerical data
- Do we have missing values?
- Do we have duplicates?

In [3]:
apple.columns

Index(['       State', '1982', '1987', '1992', '1997', '2002', '2007',
       '  1982', '  1987', '1992.1', '1997.1', '2002.1', '2007.1', '2002.2',
       '2007.2', '2002.3', '2007.3'],
      dtype='object')

In [4]:
apple.dtypes

       State    object
1982            object
1987            object
1992            object
1997            object
2002            object
2007            object
  1982          object
  1987          object
1992.1          object
1997.1          object
2002.1          object
2007.1          object
2002.2          object
2007.2          object
2002.3          object
2007.3          object
dtype: object

In [5]:
apple.shape

(50, 17)

In [6]:
apple['       State'].unique

<bound method Series.unique of 0               Alabama
1                Alaska
2               Arizona
3              Arkansas
4            California
5              Colorado
6           Connecticut
7              Delaware
8               Florida
9               Georgia
10               Hawaii
11                Idaho
12             Illinois
13              Indiana
14                 Iowa
15               Kansas
16             Kentucky
17            Louisiana
18                Maine
19             Maryland
20        Massachusetts
21             Michigan
22            Minnesota
23          Mississippi
24             Missouri
25              Montana
26             Nebraska
27               Nevada
28        New Hampshire
29           New Jersey
30           New Mexico
31             New York
32       North Carolina
33         North Dakota
34                 Ohio
35             Oklahoma
36               Oregon
37         Pennsylvania
38         Rhode Island
39       South Carolina
40       

In [7]:
apple.describe()

State 1982 1987  1992  1997  2002 2007   1982   1987 1992.1  \
count            50   50   50    50    50    50   50     50     50     50   
unique           50   49   48    49    49    50   49     49     49     49   
top         Alabama   --   --  392   485   265   35      --     --    (D)   
freq              1    2    2     2     2     1    2      2      2      2   

        1997.1 2002.1 2007.1 2002.2 2007.2 2002.3 2007.3  
count       50     50     50     50     50     50     50  
unique      48     48     48     46     49     46     48  
top     5,389     (D)    (D)      8    427    (D)    (D)  
freq         2      3      3      2      2      5      3

In [9]:
apple.isna().sum()

       State    0
1982            0
1987            0
1992            0
1997            0
2002            0
2007            0
  1982          0
  1987          0
1992.1          0
1997.1          0
2002.1          0
2007.1          0
2002.2          0
2007.2          0
2002.3          0
2007.3          0
dtype: int64

In [10]:
apple.duplicated().sum()

0

What needs cleaning with the data?

### Clean the data

Re-read the file and make sure we have get the missing values correct.

In [11]:
apple=pd.read_csv('./data/apple_data.tsv', delimiter='\t', na_values=['--','(D)'])
apple.head()

State    1982    1987    1992    1997    2002    2007     1982  \
0        Alabama    899     626     474     479     265     231    1,548    
1         Alaska     NaN     NaN      2       4       9      10       NaN   
2        Arizona    214     224     235     158     185     496    1,969    
3       Arkansas    512     451     247     238     164     153    2,276    
4     California  2,904   2,787   3,082   3,454   2,120   2,074   29,933    

      1987   1992.1   1997.1   2002.1   2007.1 2002.2 2007.2  2002.3  2007.3  
0   1,217    1,011      968      594      393     191    144     386     307  
1      NaN      NaN       2       13       15       8     10     NaN     NaN  
2   4,881    5,770    4,003    1,581    1,344     118    424   1,459   1,249  
3   2,120    1,430    1,048      658      287     123    116     560     220  
4  34,105   42,024   49,704   38,268   22,184   1,850  1,850  35,856  20,954

Make the column names correct

In [54]:
apple2 = apple.set_axis(['State', '1982_Farms', '1987_Farms', '1992_Farms', '1997_Farms', 
                         '2002_Farms', '2007_Farms', '1982_TTLAcres', '1987_TTLAcres', '1992_TTLAcres',
                         '1997_TTLAcres', '2002_TTLAcres', '2007_TTLAcres', '2002_FarmBear_Acres', 
                         '2007_FarmBear_Acres', '2002_Bear_Acres', '2007_Bear_Acres' ], axis=1)
apple2.head()

State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California     2,904      2,787      3,082      3,454      2,120    

  2007_Farms 1982_TTLAcres 1987_TTLAcres 1992_TTLAcres 1997_TTLAcres  \
0       231         1,548         1,217         1,011           968    
1        10            NaN           NaN           NaN            2    
2       496         1,969         4,881         5,770         4,003    
3       153         2,276         2,120         1,430         1,048    
4     2,074        29,933        34,105        42,024        49,704    

  2002_TTLAcres 2007_TTLAcres 2002_FarmBear_Acres 2007_FarmBear_Acres  \
0          594           393                  191                 144   
1           13            15                    8                  10   
2        1,581         1,344                  118                 424   
3          658           287                  123                 116   
4       38,268        22,184                1,850               1,850   

  2002_Bear_Acres 2007_Bear_Acres  
0             386             307  
1             NaN             NaN  
2           1,459           1,249  
3             560             220  
4          35,856          20,954

Next, remove commas. We need to alter the strings of multiple columns. We know how to do this within a single column, but how do we do this with multiple columns?  Remember the `apply()` function! We can apply any function we want.  We can't use the `pd.Series.str.replace()` function though because that belongs to the Pandas Series objects, we need an independent function. We can write our own!

We can use the formal defintion (e.g.):

```python
def remoteCommas(col):
    return col.str.replace(',','')

apple_data = apple_data.apply(remoteCommas)
```

Or we can simplify it using Pythons `lambda` function definition. `lambda` is used to declare an in-line function.  It works just like a normal function but it does not have a name and is used for quick very short in-line functions.  For example:

```python
apple_data = apple_data.apply(lambda x: x.str.replace(',',''))
```

In [55]:
apple2 = apple2.apply(lambda x: x.str.replace(',',''))
apple2.head()

State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California      2904       2787       3082       3454       2120    

  2007_Farms 1982_TTLAcres 1987_TTLAcres 1992_TTLAcres 1997_TTLAcres  \
0       231          1548          1217          1011           968    
1        10            NaN           NaN           NaN            2    
2       496          1969          4881          5770          4003    
3       153          2276          2120          1430          1048    
4      2074         29933         34105         42024         49704    

  2002_TTLAcres 2007_TTLAcres 2002_FarmBear_Acres 2007_FarmBear_Acres  \
0          594           393                  191                 144   
1           13            15                    8                  10   
2         1581          1344                  118                 424   
3          658           287                  123                 116   
4        38268         22184                 1850                1850   

  2002_Bear_Acres 2007_Bear_Acres  
0             386             307  
1             NaN             NaN  
2            1459            1249  
3             560             220  
4           35856           20954

Remove spaces preceeding the state names

In [56]:
apple2['State'].str.replace('   ', '').unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina',
       'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype=object)

Convert columns to numeric. Convert to floats instead of integers because NaNs are considered float values.

In [60]:
temp = apple2.iloc[:,1:].apply(lambda x: x.astype('float'))
apple2 = pd.concat([apple2['State'], temp], axis=1)
apple2.dtypes

State                   object
1982_Farms             float64
1987_Farms             float64
1992_Farms             float64
1997_Farms             float64
2002_Farms             float64
2007_Farms             float64
1982_TTLAcres          float64
1987_TTLAcres          float64
1992_TTLAcres          float64
1997_TTLAcres          float64
2002_TTLAcres          float64
2007_TTLAcres          float64
2002_FarmBear_Acres    float64
2007_FarmBear_Acres    float64
2002_Bear_Acres        float64
2007_Bear_Acres        float64
dtype: object

Check the types to make sure all is good

Now desribe the numeric data

### Explore the Data Again

Now that we have problems with the dataframe correct, let's explore. Are our data types correct?

Now that we have numeric data we can perform a proper summary

Do we have any missing values?

Are missing values reasonable?  Should we remove any?

Do we have any duplicated data?

Does our only categorical column have expected values?

How many rows do we have per state? 

Are the number of rows per state what we expect?

### Tidy the data

What is/are the observable unit(s)?

What are the variables?

What are the observations (values)?

In [41]:
apple2.head()

1987_Farms  1992_Farms  1997_Farms  2002_Farms  2007_Farms  1982_TTLAcres  \
0       626.0       474.0       479.0       265.0       231.0         1548.0   
1         NaN         2.0         4.0         9.0        10.0            NaN   
2       224.0       235.0       158.0       185.0       496.0         1969.0   
3       451.0       247.0       238.0       164.0       153.0         2276.0   
4      2787.0      3082.0      3454.0      2120.0      2074.0        29933.0   

   1987_TTLAcres  1992_TTLAcres  1997_TTLAcres  2002_TTLAcres  2007_TTLAcres  \
0         1217.0         1011.0          968.0          594.0          393.0   
1            NaN            NaN            2.0           13.0           15.0   
2         4881.0         5770.0         4003.0         1581.0         1344.0   
3         2120.0         1430.0         1048.0          658.0          287.0   
4        34105.0        42024.0        49704.0        38268.0        22184.0   

   2002_FarmBear_Acres  2007_FarmBear_Acres  2002_Bear_Acres  2007_Bear_Acres  
0                191.0                144.0            386.0            307.0  
1                  8.0                 10.0              NaN              NaN  
2                118.0                424.0           1459.0           1249.0  
3                123.0                116.0            560.0            220.0  
4               1850.0               1850.0          35856.0          20954.0

Next, tidy the data by melting the data. We have multiple variables in the header and we must melt them individually.  The best way to do this is to melt each variable separately, then combine them back to a single data frame. 

Do this, you will need to perform multiple melts:
1. First melt only the colums that have the number of farms and name the column `Farms`.
2. Second, melt only the columns that have the number of acres and name the column `Acres`.
3. Third, melt only the columns that have the number of bearing farms and name the column `Bearing_Farms`,
4. Fourth, melt only the columns that have the number of bearing acres and name the column `Bearing_Acres`.

You'll also want to remove the prefix (e.g. "Farms-") from each value after melting and convert the 'Year' column to an integer.

For example, after the first melt you should have a dataframe that looks like the following:

|  | State | Year | Farms |
| - | ----- | ---- | --------- |
| 0 | Alabama | 1982 | 899 |
| 1 | Alaska | 1982 | NaN |
| 2 | Arizona | 1982 | 214 |
| 3 | Arkansas | 1982 | 512 |
| 4 | California | 1982 | 2904 |

Melt #1

In [61]:
apple_melt1 = pd.melt(apple2, id_vars='State', var_name='Year_VarName',
                      value_name='Farms', ignore_index=True)
apple_melt1.head()

State Year_VarName   Farms
0        Alabama   1982_Farms   899.0
1         Alaska   1982_Farms     NaN
2        Arizona   1982_Farms   214.0
3       Arkansas   1982_Farms   512.0
4     California   1982_Farms  2904.0

Melt #2

In [63]:
apple_melt1['Year_VarName'].str.split('_', n=1, expand=True)


0           1
0    1982       Farms
1    1982       Farms
2    1982       Farms
3    1982       Farms
4    1982       Farms
..    ...         ...
795  2007  Bear_Acres
796  2007  Bear_Acres
797  2007  Bear_Acres
798  2007  Bear_Acres
799  2007  Bear_Acres

[800 rows x 2 columns]

Melt #3

Melt #4

Now combine all four melted dataframes into a single dataframe. To do this we can use the [pd.DataFrame.join()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html) function. We will need to join one table at a time.  

Note: when you join tables they must have the same indexes. So, be sure to set the index of each dataframe before joining by using the [set_index()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html) function.

In [ ]:
# Set the indexes to be the same


In [ ]:
# Join all of the tables.


What's the shape of the data now?

Do we have any duplicated rows now?

Now, do we want to remove missing values?

### Explore the Data Again

Now that we have our data fully cleaned and tidy'ed let's explore it to make sure all looks good. First, we should still have 50 states. Can you find out?

How many years do we have per state? We should have the same number except for the states where we removed rows with missing values

Describe the data, but make sure it does so by state.